### PR TITLE
convert cv8.c to D, add varstats.d

### PR DESCRIPTION
**Pull request by WalterBright**
A reboot of https://github.com/dlang/dmd/pull/8476

### CODE REVIEW

File: src/dmd/backend/varstats.d [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+ * Authors:     Rainer Schuetze
+ * License:     Distributed under the Boost Software License, Version 1.0.
+ *              http://www.boost.org/LICENSE_1_0.txt
+ * Source:      https://github.com/dlang/dmd/blob/master/src/dmd/backend/varstats.d
+ */
+
+module dmd.backend.varstats;
+
+/******************************************
+ * support for lexical scope of local variables
```
Comment (Geod24) [inline src/dmd/backend/varstats.d]: Make it part of the module DDOC

File: src/dmd/backend/varstats.d [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+ */
+
+import dmd.backend.cc;
+import dmd.backend.cdef;
+
+extern (C++):
+
+// estimate of variable life time
```
Comment (Geod24) [inline src/dmd/backend/varstats.d]: DDOC

File: src/dmd/backend/varstats.d [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+struct LifeTime
+{
+    Symbol* sym;
+    int offCreate;  // variable created before this code offset
+    int offDestroy; // variable destroyed after this code offset
```
Comment (Geod24) [inline src/dmd/backend/varstats.d]: DDOC

File: src/dmd/backend/varstats.d [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+struct LineOffset
+{
+    targ_size_t offset;
+    uint linnum;
+    uint diffNextOffset;
+}
+
+struct VarStatistics
+{
+    //this();
```
Comment (Geod24) [inline src/dmd/backend/varstats.d]: Remove ?

File: src/dmd/backend/cv8.d [hunk 5]
```diff
@@ -5,65 +5,82 @@
...
-typedef unsigned char ubyte;
-typedef unsigned short ushort;
-#define mask(m) (1 << (m))
-
-#if MARS
-#if TARGET_WINDOS
+module dmd.backend.cv8;
+
+version (MARS)
+{
```
Comment (Geod24) [inline src/dmd/backend/cv8.d]: This covers the whole file, doesn't it ?

File: src/dmd/backend/cv8.d [hunk 6]
```diff
@@ -5,65 +5,82 @@
...
+
+import core.stdc.stdio;
+import core.stdc.stdlib;
+import core.stdc.string;
+extern (C) char* getcwd(char*, size_t);
```
Comment (Geod24) [inline src/dmd/backend/cv8.d]: `import core.stdc.unistd` ?

File: src/dmd/backend/cv8.d [hunk 7]
```diff
@@ -5,65 +5,82 @@
...
+static if (TARGET_WINDOS)
+{
+
+int REGSIZE();
+extern (D) uint mask(uint m); //{ return 1 << m; }
+
+// Determine if this Symbol is stored in a COMDAT
+bool symbol_iscomdat(Symbol* s)
+{
+    version (MARS)
```
Comment (Geod24) [inline src/dmd/backend/cv8.d]: Isn't this redundant given this is already in `version (MARS)` ?

File: src/dmd/backend/cv8.d [hunk 8]
```diff
@@ -83,48 +100,45 @@ struct FuncData
...
 
-static Outbuffer *funcdata;     // array of FuncData's
+__gshared FuncData currentfuncdata;
 
-static Outbuffer *linepair;     // array of offset/line pairs
+private __gshared Outbuffer *funcdata;     // array of FuncData's
 
-uint cv8_addfile(const char *filename);
-void cv8_writesection(int seg, uint type, Outbuffer *buf);
+private __gshared Outbuffer *linepair;     // array of offset/line pairs
```
Comment (Geod24) [inline src/dmd/backend/cv8.d]: Nit: `Outbuffer* linepair`

File: src/dmd/backend/cv8.d [hunk 9]
```diff
@@ -133,121 +147,150 @@ void cv8_writename(Outbuffer *buf, const char* name, size_t len)
...
 {
     //printf("cv8_initfile()\n");
 
     // Recycle buffers; much faster than delete/renew
 
     if (!F1_buf)
-        F1_buf = new Outbuffer(1024);
-    F1_buf->setsize(0);
+    {
+        __gshared Outbuffer f1buf;
```
Comment (Geod24) [inline src/dmd/backend/cv8.d]: Why did you move away from `new` ? There's a lot of redundant code here.

File: src/dmd/backend/cv8.d [hunk 10]
```diff
@@ -639,73 +685,78 @@ void cv8_outsym(Symbol *s)
...
+            buf.writeWordn(0x1111);
+            buf.write32(cast(uint)(s.Soffset + base + BPoff));
+            buf.write32(typidx);
+            buf.writeWordn(I64 ? 334 : 22);       // relative to RBP/EBP
             cv8_writename(buf, id, len);
-            buf->writeByte(0);
-#else
+            buf.writeByte(0);
+}
+else
```
Comment (Geod24) [inline src/dmd/backend/cv8.d]: Should we keep this dead code ?

File: src/dmd/backend/varstats.d (hunk 1 above)
Comment (WalterBright) [reply src/dmd/backend/varstats.d]: First translate, then fix/refactor.

File: src/dmd/backend/varstats.d (hunk 2 above)
Comment (WalterBright) [reply src/dmd/backend/varstats.d]: First translate, then fix/refactor

File: src/dmd/backend/cv8.d (hunk 5 above)
Comment (WalterBright) [reply src/dmd/backend/cv8.d]: yes

File: src/dmd/backend/varstats.d (hunk 4 above)
Comment (WalterBright) [reply src/dmd/backend/varstats.d]: first translate, then fix/refactor

File: src/dmd/backend/cv8.d (hunk 6 above)
Comment (WalterBright) [reply src/dmd/backend/cv8.d]: no such file

File: src/dmd/backend/cv8.d (hunk 7 above)
Comment (WalterBright) [reply src/dmd/backend/cv8.d]: Yes, but there are problems with where it should go. Needs a refactoring, but first translate, then fix/refactor.

File: src/dmd/backend/cv8.d (hunk 8 above)
Comment (WalterBright) [reply src/dmd/backend/cv8.d]: first translate, then fix/refactor

File: src/dmd/backend/cv8.d (hunk 9 above)
Comment (WalterBright) [reply src/dmd/backend/cv8.d]: because `new` doesn't work in betterC, and it's never free'd anyway.

File: src/dmd/backend/cv8.d (hunk 10 above)
Comment (WalterBright) [reply src/dmd/backend/cv8.d]: First translate, then fix/refactor.

File: src/dmd/backend/cv8.d (hunk 5 above)
Comment (Geod24) [reply src/dmd/backend/cv8.d]: Then `version (MARS):` would be a better transaltion

**Comment by WalterBright** [comment]
I've said this many times, and will say it again. Translating code is fraught with difficulties and bugs that get introduced. It can be very hard to track down what went wrong. I've found through decades of experience that the way to minimize the problems is minimize the diffs, with the idea being the fewer diffs, the fewer places to look for the bug.

The procedure is:

1. prep the file for conversion, i.e. try to minimize the use of idioms that won't easily translate
2. pass the test suite
3. translate the code with as few edits as practical. Do not reformat the code. Do not refactor it. Do not fix anything, no matter how tempting. Reproduce the behavior of the original as much as possible.
4. pass the test suite
5. now reformat, refactor, fix (as separate PRs, of course, and passing the test suite in between each).
6. pass the test suite

I learned this the hard way, many times over, because I'm a slow learner.

File: src/dmd/backend/cv8.d (hunk 6 above)
Comment (Geod24) [reply src/dmd/backend/cv8.d]: Right, that is [`core.sys.posix.unistd`](https://github.com/dlang/druntime/search?q=getcwd&unscoped_q=getcwd)

Comment (WalterBright) [reply src/dmd/backend/cv8.d]: Windows isn't Posix, and the whole file is version'd for `Posix`. Dealing with these issues is beyond the scope of this PR, and won't work anyway because DMD is compiled with an older version of DMD that would retain these issues.

**Comment by Geod24** [comment]
@WalterBright : I agree with most of that. Didn't see the changes requested as things that would break with that though.

**Review by Geod24 — APPROVED** [review]

File: src/dmd/backend/cv8.d [hunk 11]
```diff
@@ -514,23 +559,23 @@ uint cv8_addfile(const char *filename)
...
-    ubyte *p = F3_buf->buf;
+    uint length = cast(uint)F3_buf.size();
+    ubyte *p = F3_buf.buf;
     size_t len = strlen(filename);
 
     // ensure the filename is absolute to help the debugger to find the source
     // without having to know the working directory during compilation
-    static char cwd[260];
-    static uint cwdlen;
+    __gshared char[256] cwd;
```
Comment (rainers) [inline src/dmd/backend/cv8.d]: Why change the size here? MAX_PATH is defined as 260.

File: src/dmd/backend/cv8.d [hunk 12]
```diff
@@ -514,23 +559,23 @@ uint cv8_addfile(const char *filename)
...
+    __gshared uint cwdlen;
     bool abs = (*filename == '\\') ||
                (*filename == '/')  ||
                (*filename && filename[1] == ':');
 
     if (!abs && cwd[0] == 0)
```
Comment (rainers) [inline src/dmd/backend/cv8.d]: This won't work anymore as `char` is initialized to 0xff, not 0.

File: src/dmd/backend/cv8.d [hunk 13]
```diff
@@ -639,73 +684,78 @@ void cv8_outsym(Symbol *s)
...
+            {   base = cast(uint)Fast.size;
                 goto L1;
             }
             goto L2;
 
         case SCregister:
-            if (s->Sfl != FLreg)
+            if (s.Sfl != FLreg)
                 goto case_auto;
+            goto L2;
```
Comment (rainers) [inline src/dmd/backend/cv8.d]: nit: `goto case;` would be a closer translation.

**Review by rainers — DISMISSED** [review]

File: src/dmd/backend/cv8.d (hunk 11 above)
Comment (WalterBright) [reply src/dmd/backend/cv8.d]: A stupid mistake of mine.

File: src/dmd/backend/cv8.d (hunk 12 above)
Comment (WalterBright) [reply src/dmd/backend/cv8.d]: good catch

File: src/dmd/backend/cv8.d (hunk 13 above)
Comment (WalterBright) [reply src/dmd/backend/cv8.d]: ok

File: src/dmd/backend/cv8.d [hunk 14]
```diff
@@ -5,65 +5,81 @@
...
+
+extern (C++):
+
+static if (TARGET_WINDOS)
+{
+
+int REGSIZE();
+
+// Determine if this Symbol is stored in a COMDAT
+bool symbol_iscomdat(Symbol* s)
```
Comment (rainers) [inline src/dmd/backend/cv8.d]: Why is symbol_iscomdat defined here, if it already exists (as comment from C++ conversion of cc.h) in cc.d? It is needed elsewhere, too.

Comment (WalterBright) [reply src/dmd/backend/cv8.d]: Because I'm doing as little as possible to translate the code to D. Refactoring comes later.

Comment (rainers) [reply src/dmd/backend/cv8.d]: I'd consider copying the function from elsewhere more refactoring than leaving it in place but YMMV.

**Comment by jacob-carlborg** [comment]
> The multiple commits can be squashed when merging (@wilzbach can this be done automatically with auto-merge?).

We should only do this for fixup commits.

**Comment by rainers** [comment]
> Is the semaphoreci failure real or spurious? I couldn't find anything comprehensible in the log.

It seems to have timed out after an hour. As cv8 is unlikely to affect a linux build, I'd consider it spurious. semaphoreci isn't set to be required, so why isn't this merging with the auto-merge label being set?

**Comment by rainers** [comment]
> > The multiple commits can be squashed when merging (@wilzbach can this be done automatically with auto-merge?).
> 
> We should only do this for fixup commits.

I was thinking of an additional/alternative label `auto-squash` that would have to be set.

**Comment by Geod24** [comment]
Auto-merge toggled on
